libraries/pet_service/pet_service.lua

```diff
@@ -0,0 +1,176 @@
+            data.hp_percent = p.hp_percent
+            data.mp_percent = p.mp_percent
+            data.tp = p.pet_tp
+            if p.pet_index and p.pet_index ~= 0 then
+                data.active = true
+            else
+                data.active = false
+            end
+        end
+    end,
+    [{0x068}] = function(p)
+        if p.type == 4 then
+            data.owner_index = p.owner_index
+            data.owner_id = p.owner_id
+            data.index = p.pet_index
+            data.hp_percent = p.hp_percent
+            data.mp_percent = p.mp_percent
+            data.tp = p.pet_tp
+            data.target_id = p.target_id
+            data.name = p.pet_name
+            if p.pet_index and p.pet_index ~= 0 then
+                data.active = true
+            else
+                data.active = false
+            end
+        end
+    end,
+    [{0x044,0x12}] = function(p)
+
+        data.automaton.head.raw         = p.automaton_head
+        data.automaton.head.item_id     = p.automaton_head + 0x2000
+        data.automaton.frame.raw        = p.automaton_frame
+        data.automaton.frame.item_id    = p.automaton_frame + 0x2000
+        for i=0, 11 do
+            data.automaton.attachments[i].raw = p.attachments[i]
+            data.automaton.attachments[i].item_id = p.attachments[i] + 0x2100
+        end
+        ffi.copy(data.automaton._available_heads, p._available_heads, 4)
+        ffi.copy(data.automaton._available_frames, p._available_frames, 4)
+        ffi.copy(data.automaton._available_attach, p._available_attach, 32)
+        data.automaton.name             = p.pet_name
+        data.automaton.hp               = p.hp
+        data.automaton.hp_max           = p.hp_max
+        data.automaton.mp               = p.mp
+        data.automaton.mp_max           = p.mp_max
+        data.automaton.melee            = p.melee
+        data.automaton.melee_max        = p.melee_max
+        data.automaton.ranged           = p.ranged
+        data.automaton.ranged_max       = p.ranged_max
+        data.automaton.magic            = p.magic
+        data.automaton.magic_max        = p.magic_max
+        data.automaton.str              = p.str
+        data.automaton.str_modifier     = p.str_modifier
+        data.automaton.dex              = p.dex
+        data.automaton.dex_modifier     = p.dex_modifier
+        data.automaton.vit              = p.vit
+        data.automaton.vit_modifier     = p.vit_modifier
+        data.automaton.agi              = p.agi
+        data.automaton.agi_modifier     = p.agi_modifier
+        data.automaton.int              = p.int
+        data.automaton.int_modifier     = p.int_modifier
+        data.automaton.mnd              = p.mnd
+        data.automaton.mnd_modifier     = p.mnd_modifier
+        data.automaton.chr              = p.chr
+        data.automaton.chr_modifier     = p.chr_modifier
+
+        local active = data.active and (data.name == data.automaton.name)
```

Comment:
I haven't examined this myself, but it feels a bit brittle. Is that the best way to check for an active pet? Can `data.automaton.name` ever be something other than `data.name` or empty? If we switch to a pointer based approach this should be moved to the top and the function should just return if it's `false`.

Reply:
This compares the automaton's name (`pet.automaton.name`) to the active pet's name (`pet.name`) to tell if the automaton is the active pet or not.

This is the only way I managed to think of to check this given the known packet fields, but there might be a better way to tell that we just haven't exposed yet. The client somehow knows the automaton is active and will say "You cannot customize an active automaton." if you try to change attachments. I'm skeptical SE actually left this up to something like comparing the name, but I've looked at the known pet relate packets and I don't see anything else in them that would tell the client the active pet is the PUP automaton...